### ISSUE TITLE
Support indices creation from the package

### ISSUE THREAD

**Issue by aleksmaus** [elastic/package-spec#117]
There is a use case specifically with Fleet Server integration package where we would prefer to have not only indices templates creates, but the indices itself. 

Two considerations:
1. kibana_system presently doesn't have permissions to create indices
2. we probably should not fail the package install if the index exists. what are the possible issues with that?

**Comment by ruflin** [comment]
The way I was thinking of this feature is that under `elasticsearch/index/{index-name}.json` a json object could be stored which creates the index. This uses the Index Create API: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-create-index.html#indices-create-index

In the simplest case, the json could be empty, more complex cases can contain mappings / settings etc.

What happens on upgrade / deletion of a package?

**Comment by ycombinator** [comment]
I have a slightly different suggestion that might handle the update use case better.

The index definitions are stored under `elasticsearch/index/{index-name}/N` folder, where `N` corresponds to the migration number, 0-based. This migration number is stored in the [index's `_meta` field](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-meta-field.html).

So the initial index definition goes into `elasticsearch/index/{index-name}/0`, the next one will go into `elasticsearch/index/{index-name}/1`, etc. When a package is installed/upgraded, Kibana checks if the index already exists.
- If it does, it queries the existing index's `_meta.version` property and applies all migrations, in sequence, greater than that queried version up to the highest one. With each successful application, the `_meta.version` property is updated to the version that was successfully applied.
- If it does not, it creates the index using the definitions in `elasticsearch/index/{index-name}/0`, while setting `_meta.version` to `0`.

Under each `elasticsearch/index/{index-name}/N` folder, there would be one or both of these files:
- `mappings.json`: contains an index mappings document, as accepted by the [Put mapping API](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html).
- `settings.json`: contains an index settings document, as accepted by the [Update settings API](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-update-settings.html).

Separating the settings and mappings into their own files is mainly done as a matter of convenience for Kibana. After the initial index creation, subsequent changes to the index definition in Elasticsearch require two separate API calls - one for settings changes and one for mappings changes - depending on the changes, of course. By defining these two change documents separately, it becomes easy for Kibana to just read them and call the appropriate API. We could merge these into a single JSON file instead (as accepted by the [Create index API](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-create-index.html)), but then Kibana would need some extra logic to know if it was creating an index - in which case it could use the entire change document as-is - or updating an index - in which case it would need to parse the document into settings and mappings sub-documents and then call the respective update APIs.

I'm not sure what should happen when a package is deleted. I guess that depends on whether it's desirable to keep the data indexed by the package or not. My instinct says we should not delete data, i.e. we should leave the index as-is, but I don't know enough about the use case for this feature to have a solid opinion on this aspect of the feature.

**Comment by ruflin** [comment]
What if there is a conflict of the previous to the new mapping? Now that you brought up versions, I wonder if we should do something similar / same as the `.kibana` index. The old trick of using an alias and create an index behind it. If the mapping is not compatible anymore, we create a new one and set it as the new write index on the alias. Having the alias also ensures, we can always read from all the data in an easy way.

But I'm not sure if this would just fall under `elasticsearch/index` as it is a much more complex feature. I kind of hoped with data streams we could stay away from this but not everything should be a data stream :-(

**Comment by ycombinator** [comment]
@ruflin Good point re: failure handling. ++ to the alias idea similar to how Kibana does its migrations for `.kibana`.
